File: ruby/spec/2022/day_01_spec.rb
require '2022/day_01'

# https://adventofcode.com/2022/day/1

describe "Advent 2022 Day 1" do
  context "part 1" do
    it "No elves" do
      expect(part_1("")).to eq 0
    end

    it "one elf with no calories" do
      expect(part_1("0")).to eq 0
    end

    it "one elf with one calorie" do
      expect(part_1("1")).to eq 1
    end

    it "one elf with two calories" do
      expect(part_1("2")).to eq 2
    end

    it "one elf with two rations, one being zero" do
      expect(part_1("10\n0")).to eq 10
    end

    it "one elf with two rations" do
      expect(part_1("10\n20")).to eq 30
    end

    it "two elves with one ration" do
      expect(part_1("1\n\n1")).to eq 1
    end

    it "two elves with differing rations" do
      expect(part_1("1\n\n2")).to eq 2
    end

    it "two elves with differing rations" do
      expect(part_1("1000\n2000\n3000\n\n4000\n\n5000\n6000\n\n7000\n8000\n9000\n\n10000")).to eq 24000
    end

    it "my test data" do
      expect(part_1(input_data(2022, 1))).to eq 72511
    end
  end

  context "part 2" do
    it "no elves" do
      expect(part_2("")).to eq 0
    end

    it "one elf with no calories" do
      expect(part_2("0")).to eq 0
    end

    it "one elf with one calorie" do
      expect(part_2("1")).to eq 1
    end

    it "two elves with one calorie" do
      expect(part_2("1\n\n1")).to eq 2
    end

    it "three elves with 1, 2, and 3 calories" do
      expect(part_2("1\n\n2\n\n3")).to eq 6
    end

    it "four elves with 1, 2, 3, and 4 calories" do
      expect(part_2("1\n\n2\n\n3\n\n4")).to eq 9
    end

    it "base case" do
      expect(part_2("1000\n2000\n3000\n\n4000\n\n5000\n6000\n\n7000\n8000\n9000\n\n10000")).to eq 45000
    end

    it "my test data" do
      expect(part_2(input_data(2022, 1))).to eq 212117
    end
  end
end

File: clojure/spec/aoc/2022/day01_spec.clj
(ns aoc.2022.day01-spec
  (:require [aoc.2022.day01 :refer :all]
            [aoc.util :as util]
            [speclj.core :refer :all]))

; https://adventofcode.com/2022/day/1

(def input-data (util/input-data 2022 1))
(def sample-data (util/sample-data 2022 1))

(describe "AoC Day 1"

  (context "Part 1"
    (it "no elves"
      (should= 0 (part-1 "")))

    (it "one elf with no calories"
      (should= 0 (part-1 "0")))

    (it "one elf with one calorie"
      (should= 1 (part-1 "1")))

    (it "one elf with two calories"
      (should= 2 (part-1 "2")))

    (it "two elves with one calorie"
      (should= 1 (part-1 "1\n\n1")))

    (it "two elves with different calories"
      (should= 2 (part-1 "1\n\n2")))

    (it "one elf with two rations"
      (should= 3 (part-1 "1\n2")))

    (it "solves sample data"
      (should= 24000 (part-1 sample-data)))

    (it "solves test data"
      (should= 72511 (part-1 input-data))))

  (context "Part 2"
    (it "no elves"
      (should= 0 (part-2 "")))

    (it "one elf with no calories"
      (should= 0 (part-2 "0")))

    (it "one elf with one calorie"
      (should= 1 (part-2 "1")))

    (it "one elf with two calories"
      (should= 2 (part-2 "1\n1")))

    (it "two elves"
      (should= 3 (part-2 "1\n\n2")))

    (it "three elves"
      (should= 6 (part-2 "1\n\n2\n\n3")))

    (it "four elves"
      (should= 9 (part-2 "1\n\n2\n\n3\n\n4")))

    (it "sample data"
      (should= 45000 (part-2 sample-data)))

    (it "test data"
      (should= 212117 (part-2 input-data)))))
